File: website/src/features/store-setting/data/datasources/StoreDataSource.ts
import { supabaseService } from '@/core/services/supabase_service';
import { DateTimeUtils } from '@/core/utils/datetime-utils';

/**
 * StoreDataSource - Supabase 테이블 직접 쿼리
 * Backup 방식: /backup/pages/settings/store-setting/index.html
 * RPC 함수 사용하지 않음
 */
export class StoreDataSource {
  /**
   * 회사의 모든 활성 상점 조회
   * Backup: lines 684-689
   */
  async getStores(companyId: string) {
    const { data, error } = await supabaseService
      .getClient()
      .from('stores')
      .select('store_id, store_name, store_address, store_phone, huddle_time, payment_time, allowed_distance')
      .eq('company_id', companyId)
      .eq('is_deleted', false)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(`Failed to fetch stores: ${error.message}`);
    }

    return data || [];
  }

  /**
   * 새 상점 생성
   * Backup: lines 1120-1123
   */
  async createStore(
    companyId: string,
    storeName: string,
    address: string | null,
    phone: string | null
  ) {
    const storeData = {
      company_id: companyId,
      store_name: storeName,
      store_address: address || null,
      store_phone: phone || null,
      created_at: DateTimeUtils.nowUtc(),
      updated_at: DateTimeUtils.nowUtc(),
      is_deleted: false,
    };

    const { data, error } = await supabaseService
      .getClient()
      .from('stores')
      .insert(storeData)
      .select();

    if (error) {
      throw new Error(`Failed to create store: ${error.message}`);
    }

    return data;
  }

  /**
   * 상점 삭제 (Soft Delete)
   * Backup: lines 995-1001
   */
  async deleteStore(storeId: string) {
    const { error } = await supabaseService
      .getClient()
      .from('stores')
      .update({
        is_deleted: true,
        updated_at: DateTimeUtils.nowUtc(),
      })
      .eq('store_id', storeId);

    if (error) {
      throw new Error(`Failed to delete store: ${error.message}`);
    }
  }
}

File: myFinance_improved_V2/lib/features/store_shift/data/datasources/store_datasource.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../../../core/services/supabase_service.dart';
import '../../../../core/utils/datetime_utils.dart';
import '../../domain/exceptions/store_shift_exceptions.dart';

/// Store Data Source
///
/// Handles Supabase operations for store information and settings
class StoreDataSource {
  final SupabaseService _supabaseService;

  StoreDataSource(this._supabaseService);

  SupabaseClient get _client => _supabaseService.client;

  /// Get detailed store information by ID
  ///
  /// Uses RPC function 'get_store_info_v2' to fetch store with latitude/longitude
  Future<Map<String, dynamic>> getStoreById({
    required String storeId,
    required String companyId,
  }) async {
    try {
      final response = await _client.rpc<List<dynamic>>(
        'get_store_info_v2',
        params: {'p_company_id': companyId},
      );

      // Find the store by storeId from the list
      final store = response.cast<Map<String, dynamic>>().firstWhere(
        (s) => s['store_id'] == storeId,
        orElse: () => <String, dynamic>{},
      );

      if (store.isEmpty) {
        throw Exception('Store not found');
      }

      return store;
    } catch (e, stackTrace) {
      throw StoreNotFoundException(
        'Failed to fetch store $storeId: $e',
        stackTrace,
      );
    }
  }

  /// Update store location
  ///
  /// Uses RPC function 'update_store_location'
  Future<void> updateStoreLocation({
    required String storeId,
    required double latitude,
    required double longitude,
    required String address,
  }) async {
    try {
      await _client.rpc<void>(
        'update_store_location',
        params: {
          'p_store_id': storeId,
          'p_latitude': latitude,
          'p_longitude': longitude,
          'p_address': address,
        },
      );
    } catch (e, stackTrace) {
      throw StoreLocationUpdateException(
        'Failed to update store location: $e',
        stackTrace,
      );
    }
  }

  /// Update operational settings
  ///
  /// Uses RPC function 'update_store_setting'
  Future<void> updateOperationalSettings({
    required String storeId,
    int? huddleTime,
    int? paymentTime,
    int? allowedDistance,
  }) async {
    try {
      final localTime = DateTimeUtils.formatLocalTimestamp();
      final timezone = DateTimeUtils.getLocalTimezone();

      final response = await _client.rpc<Map<String, dynamic>>(
        'update_store_setting',
        params: {
          'p_store_id': storeId,
          'p_huddle_time': huddleTime,
          'p_payment_time': paymentTime,
          'p_allowed_distance': allowedDistance,
          'p_time': localTime,
          'p_timezone': timezone,
        },
      );

      if (response['success'] == false) {
        throw Exception(response['message'] ?? 'Unknown error');
      }
    } catch (e, stackTrace) {
      throw StoreLocationUpdateException(
        'Failed to update operational settings: $e',
        stackTrace,
      );
    }
  }

  /// Update store information (name, email, phone, address)
  ///
  /// Uses RPC function 'store_shift_update_store_info'
  Future<void> updateStoreInfo({
    required String storeId,
    String? storeName,
    String? storeEmail,
    String? storePhone,
    String? storeAddress,
  }) async {
    try {
      final localTime = DateTimeUtils.formatLocalTimestamp();
      final timezone = DateTimeUtils.getLocalTimezone();

      final response = await _client.rpc<Map<String, dynamic>>(
        'store_shift_update_store_info',
        params: {
          'p_store_id': storeId,
          'p_store_name': storeName,
          'p_store_email': storeEmail,
          'p_store_phone': storePhone,
          'p_store_address': storeAddress,
          'p_time': localTime,
          'p_timezone': timezone,
        },
      );

      if (response['success'] == false) {
        throw StoreLocationUpdateException(
          response['message'] as String? ?? 'Unknown error',
          StackTrace.current,
        );
      }
    } catch (e, stackTrace) {
      if (e is StoreLocationUpdateException) rethrow;

      throw StoreLocationUpdateException(
        'Failed to update store info: $e',
        stackTrace,
      );
    }
  }
}
